src/snowflake/snowpark/modin/plugin/_internal/binary_op_utils.py

```diff
@@ -312,314 +299,496 @@ def _op_is_between_timedelta_and_numeric(
-    ):
-        binary_op_result_column = cast(
-            floor(first_operand * second_operand), LongType()
-        )
-        snowpark_pandas_type = TimedeltaType()
-    # For `eq` and `ne`, note that Snowflake will consider 1 equal to
-    # Timedelta(1) because those two have the same representation in Snowflake,
-    # so we have to compare types in the client.
-    elif op == "eq" and (
-        _op_is_between_timedelta_and_numeric(first_datatype, second_datatype)
-    ):
-        binary_op_result_column = pandas_lit(False)
-    elif op == "ne" and _op_is_between_timedelta_and_numeric(
-        first_datatype, second_datatype
-    ):
-        binary_op_result_column = pandas_lit(True)
-    elif (
-        op in ("truediv", "floordiv")
-        and isinstance(first_datatype(), TimedeltaType)
-        and _is_numeric_non_timedelta_type(second_datatype())
-    ):
-        binary_op_result_column = cast(
-            floor(first_operand / second_operand), LongType()
+
+    def _get_result(self) -> SnowparkPandasColumn:
+        return SnowparkPandasColumn(
+            snowpark_column=self.result_column,
+            snowpark_pandas_type=self.result_snowpark_pandas_type,
         )
-        snowpark_pandas_type = TimedeltaType()
-    elif (
-        op == "mod"
-        and isinstance(first_datatype(), TimedeltaType)
-        and _is_numeric_non_timedelta_type(second_datatype())
-    ):
-        binary_op_result_column = ceil(
-            compute_modulo_between_snowpark_columns(
-                first_operand, first_datatype(), second_operand, second_datatype()
+
+    def _check_timedelta_with_none(self) -> None:
+        if self.op in (
+            "add",
+            "sub",
+            "eq",
+            "ne",
+            "gt",
+            "ge",
+            "lt",
+            "le",
+            "floordiv",
+            "truediv",
+        ) and (
+            (
+                isinstance(self.first_datatype(), TimedeltaType)
+                and isinstance(self.second_datatype(), NullType)
             )
+            or (
+                isinstance(self.second_datatype(), TimedeltaType)
+                and isinstance(self.first_datatype(), NullType)
+            )
+        ):
+            self.result_column = pandas_lit(None)
+            self.result_snowpark_pandas_type = TimedeltaType()
+
+    def _check_error(self) -> None:
+        if (
+            self.op == "sub"
+            and isinstance(self.first_datatype(), TimedeltaType)
+            and isinstance(self.second_datatype(), TimestampType)
+        ):
+            # Timedelta - Timestamp doesn't make sense. Raise the same error
+            # message as pandas.
+            raise TypeError("bad operand type for unary -: 'DatetimeArray'")
+
+        elif self.op in "pow" and _op_is_between_two_timedeltas_or_timedelta_and_null(
+            self.first_datatype(), self.second_datatype()
+        ):
+            raise TypeError("unsupported operand type for **: Timedelta")
```

Comment:
There seems like a like of redundancy in these error message strings, really they can be consolidated (such as a constant template string) and parameterized based on the op.

Reply:
Let me give another pass. This version was just copy and paste from the original code.